mlir/include/mlir/Dialect/NVGPU/IR/NVGPU.td

```diff
@@ -20,6 +20,7 @@
 #ifndef NVGPU
 #define NVGPU
 
+include "mlir/Interfaces/InferTypeOpInterface.td"
```

Comment:
I want to check the input/output type of `rcp` are equal with `SameOperandsAndResultType`. So you suggested we'd better change to use verifier?

Reply:
My point was about inferring the result type. When someone sees an IR like `%out = nvgpu.rcp_approx %in : vector<32x16xf32>`, the return type is unclear, making the IR unreadable. To understand the return type, one would need to look at the TableGen definition of the operation or read the C++ implementation.

```
// Current OP, unclear what is %out
%out = nvgpu.rcp_approx %in : vector<32x16xf32> 
// This is my proposal, it's clear what is %out
%out = nvgpu.rcp_approx %in : vector<32x16xf32> -> vector<32x16xf32>
```

You could use `PredOpTrait` or `SameOperandsAndResultType`, or handle it in the verifier. Does `SameOperandsAndResultType` forces to infer result type?

Reply:
Ok, got it. Let me update the assembly format. 

> You could use PredOpTrait or SameOperandsAndResultType, or handle it in the verifier. Does SameOperandsAndResultType forces to infer result type?

Yes, `SameOperandsAndResultType` relies on `InferTypeOpInterface` and it forces to create build function with `InferReturnTypes`.

Reply:
Great, `PredOpTrait` and `SameOperandsAndResultShape` successfully solved my question. Thanks a lot.

Reply:
> My point was about inferring the result type. When someone sees an IR like %out = nvgpu.rcp_approx %in : vector<32x16xf32>, the return type is unclear, making the IR unreadable. To

This is how the entirety of the arith dialect is setup: is the return type **must** be the same as the input type this should be the common idiom for formatting dialects upstream IMO.

Reply:
Agreed, the `arith` dialect handles this well, and the IR is readable. 

However, when an OP in another dialect infers the result type, I find it less readable. This is just my personal opinion.

Reply:
Thanks @joker-eph and @grypp , both of formats are good to me. Since most ops of nvgpu dialect have explicit result types, I personally think it is fine to follow this style.

Reply:
Consistency is important, so that's reasonable.

That said it's not a matter of majority: we can also see the nvgpu dialect as "not using the common upstream practices" and in need of an upgrade.

Reply:
Recovered original style. Please help check if you have other questions about this pr, thanks a lot. :-)
```
%out = nvgpu.rcp_approx %in {rounding=approx, ftz}: vector<32x16xf32>
```

Reply:
> That said it's not a matter of majority: we can also see the nvgpu dialect as "not using the common upstream practices" and in need of an upgrade.

We can upgrade it, but what's the guideline with inferring result types? It makes sense when input and output types same, like with `arith`. Otherwise, imho, it turns IR obscure and unreadable.

Reply:
>  It makes sense when input and output types same, like with arith. 

I thought this is the case we're talking about here? The example provided above was `%out = nvgpu.rcp_approx %in : vector<32x16xf32> -> vector<32x16xf32>` 

For the generality, it is slightly more subtle, for example what about `%res = memref.load %ptr[%idx] : memref<10xf32>` ?